### PR TITLE
[batch] Add kubernetes secret, service account cache

### PR DESCRIPTION
**Pull request by jigold**
This should speed up the scheduler a bit. I tested it with some log statements to ensure there were cache hits. I tested the cleanup loop works before I added a try, except wrapper.

### CODE REVIEW

File: batch/batch/driver/k8s_cache.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        sa, time_updated = self.service_accounts.get((name, namespace), (None, None))
+        if time_updated and time.time() < time_updated + self.refresh_time:
+            return sa
+
+        sa = await self.client.read_namespaced_service_account(
+            name, namespace, _request_timeout=timeout)
+        self.service_accounts[(name, namespace)] = (sa, time.time())
+        return sa
+
+    async def cleanup_loop(self):
```
Comment (cseed) [inline batch/batch/driver/k8s_cache.py]: It's probably fine, but this doesn't bound the size of the cache.  I'd probably just fix the size of the cache and remove the oldest element(s) (use a sorted dict) when adding a new element if it is full.

Comment (jigold) [reply batch/batch/driver/k8s_cache.py]: I couldn't figure out how to use sorted dict, so I implemented it with a regular dict / sorted set.

File: batch/batch/driver/k8s_cache.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        asyncio.ensure_future(self.cleanup_loop())
+
+    async def read_secret(self, name, namespace, timeout):
+        secret, time_updated = self.secrets.get((name, namespace), (None, None))
+        if time_updated and time.time() < time_updated + self.refresh_time:
+            return secret
+
+        secret = await self.client.read_namespaced_secret(
+            name, namespace, _request_timeout=timeout)
+        self.secrets[(name, namespace)] = (secret, time.time())
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/batch/driver/k8s_cache.py]: I added locking. Hopefully that solves the issue.

**Comment by jigold** [comment]
Closing in favor of #7850

**Pull request closed without merging**